src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,617 @@
+        asserts.assert_true(isinstance(result.Reason, InteractionModelError), msg="Unexpected success reading invalid endpoint")
+
+        # Seems to return only {} -- Is this a failure or the unintended behavior?
+        # read_request_19 = await self.default_controller.ReadAttribute(self.dut_node_id, [non_existent_endpoint])
+
+        # Step 20
+        # TH sends the Read Request Message to the DUT to read any attribute to an unsupported cluster.
+        # DUT responds with the report data action.
+
+        # Verify on the TH that the DUT sends the status code UNSUPPORTED_CLUSTER
+        self.print_step(20, "Send the Read Request Message to the DUT to read any attribute to an unsupported cluster")
+
+        for endpoint_id, endpoint in self.endpoints.items():
+            print(endpoint_id, endpoint)
+            for cluster_type, cluster in endpoint.items():
+                if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                    continue
+
+                all_clusters = set(list(ClusterObjects.ALL_CLUSTERS.keys()))
+                dut_clusters = set(list(x.id for x in endpoint.keys()))
+
+                unsupported = [id for id in list(all_clusters - dut_clusters) if global_attribute_ids.attribute_id_type(id)
+                               == global_attribute_ids.AttributeIdType.kStandardNonGlobal]
+
+                unsupported_attribute = (ClusterObjects.ALL_ATTRIBUTES[unsupported[0]])[0]
+
+                if unsupported:
+                    result = await self.read_single_attribute_expect_error(endpoint=endpoint_id, cluster=ClusterObjects.ALL_CLUSTERS[unsupported[0]], attribute=unsupported_attribute, error=Status.UnsupportedCluster)
+                    asserts.assert_true(isinstance(result.Reason, InteractionModelError),
+                                        msg="Unexpected success reading invalid cluster")
+
+        # Step 21
+        # TH sends the Read Request Message to the DUT to read an unsupported attribute
+        # DUT responds with the report data action.
+        self.print_step(21, "Send the Read Request Message to the DUT to read any attribute to an unsupported attribute")
+
+        for endpoint_id, endpoint in self.endpoints.items():
+            for cluster_type, cluster in endpoint.items():
+                if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                    continue
+
+            all_attrs = set(list(ClusterObjects.ALL_ATTRIBUTES[cluster_type.id].keys()))
+            dut_attrs = set(cluster[cluster_type.Attributes.AttributeList])
+
+            unsupported = [id for id in list(all_attrs - dut_attrs) if global_attribute_ids.attribute_id_type(id)
+                           == global_attribute_ids.AttributeIdType.kStandardNonGlobal]
+
+            if unsupported:
```

Comment:
please add a break after this is done - only need to do one cluster.

Reply:
Will appear in next commit